### ISSUE TITLE
TaskT_SyncAsync_Extensions.IterT is eager

### ISSUE THREAD

**Issue by chyczewski-maciej** [louthy/language-ext#620]
`IEnumerable<Task<T>>.IterT` extension method is eager. Is it expected? Is this related to #564?

This can lead to unintentional iteration (in some cases creation and start) of all tasks in a sequence.
Here's a small snippet that'd show an extreme case of the issue. In less extreme cases it may lead to uncontrolled parallelism which goes unnoticed.

```
public class Program
{
    public static void ConsoleLog(string str) => Console.WriteLine($"{DateTime.Now.ToLongTimeString()}: {str}");

    public static IEnumerable<Task<int>> GetNumbersAsync()
    {
        int i = 0;
        while (true) // Mimic infinite (or just very long sequence of data)
            yield return GetNumberAsync();

        async Task<int> GetNumberAsync()
        {
            var value = Interlocked.Increment(ref i);
            if (value % 100_000 == 0) // I'm not printing every single task to avoid spamming the screenshot and to show how many tasks can be created in just a few seconds
                ConsoleLog($"Starting {nameof(GetNumberAsync)} Task #{value}");

            await Task.Delay(1000); // Mimic some network/disk IO
            return value;
        }
    }

    public static async Task Main(string[] args)
    {
        ConsoleLog("Starting the application");
        await GetNumbersAsync()
                .IterT(value =>
                {
                    ConsoleLog($"Value: {value}");
                    Thread.Sleep(500); // Mimic some CPU intensive work
                });
    }
}
```
![image](https://user-images.githubusercontent.com/17740400/63032310-dac1c200-beb5-11e9-8267-f55bda9c412e.png)
![image](https://user-images.githubusercontent.com/17740400/63033177-7ef83880-beb7-11e9-8101-bca245bd2a64.png)

Because the enumeration of the GetNumbersAsync is eager it'll create hundreds of thousands of tasks that start right away in the background. If GetNumbersAsync would query some service instead we would bomb the service with hundreds of thousands requests.

In this example application managed to spawn 3.2kk tasks while it was able to process only 10 values.

More information on the extension method:
namespace: `LanguageExt`
class: `TaskT_SyncAsync_Extensions`
method signature: `public static Task<Unit> IterT<A>(this IEnumerable<Task<A>> ma, Action<A> f)`

**Comment by louthy** [comment]
`Iter` and `IterT` are always eager.  Their whole purpose is to iterate the sequence by pulling on it.

**Comment by chyczewski-maciej** [comment]
@louthy, thanks for the response. 
I may have badly described the issue. Enumeration of `IEnumerable<Task<T>>` is done in **parallel** and independently of **sequentially** executed action for each element.
This may lead to enumeration of elements that are not needed at a time.

Let me use some (oversimplified) examples to explain myself.
This is the kind of I'd expect.
Here we enumerate only the elements that are currently needed for the execution.
```
foreach(var e in elements)
  action(e);
```

This is the actual behavior:
Here all elements are enumerated regardless of when they are needed.
```
foreach(var e in elements.ToList())
   action(e);
```

**Comment by bender2k14** [comment]
I also don't understand what you mean @chyczewski-maciej.

I suppose we `Take(10)` from the `IEnumerable<Task<int>>` after mapping each integer to its square.

```
public class TestClass
{
    private readonly ITestOutputHelper _output;

    public TestClass(ITestOutputHelper output)
    {
        _output = output;
    }

    [Fact]
    public async Task Method()
    {
        Log("Starting the application");
        await GetNumbersAsync()
            .MapT(i => i * i)
            .Take(10)
            .IterT(value =>
            {
                Log($"Value: {value}");
                Thread.Sleep(500); // Mimic some CPU intensive work
            });
    }


    private void Log(string str) => _output.WriteLine($"{DateTime.Now.ToLongTimeString()}: {str}");

    private IEnumerable<Task<int>> GetNumbersAsync()
    {
        var i = 0;
        while (true) // Mimic infinite (or just very long sequence of data)
            yield return GetNumberAsync();

        async Task<int> GetNumberAsync()
        {
            var value = Interlocked.Increment(ref i);
            Log($"Starting {nameof(GetNumberAsync)} Task #{value}");

            await Task.Delay(1000); // Mimic some network/disk IO
            return value;
        }
    }
}
```

How many lines do you expect this to log?  Of the log entries from `Log($"Starting {nameof(GetNumberAsync)} Task #{value}");`, what values do you expect will appear?

When I run this test, there are 20 lines in the log and the values from that particular line are the numbers 1 to 10.  That is exactly what I expect to happen.  Is that what you expected to happen?

```
1:48:05 PM: Starting the application
1:48:05 PM: Starting GetNumberAsync Task #1
1:48:05 PM: Starting GetNumberAsync Task #2
1:48:05 PM: Starting GetNumberAsync Task #3
1:48:05 PM: Starting GetNumberAsync Task #4
1:48:05 PM: Starting GetNumberAsync Task #5
1:48:05 PM: Starting GetNumberAsync Task #6
1:48:05 PM: Starting GetNumberAsync Task #7
1:48:05 PM: Starting GetNumberAsync Task #8
1:48:05 PM: Starting GetNumberAsync Task #9
1:48:05 PM: Starting GetNumberAsync Task #10
1:48:06 PM: Value: 1
1:48:06 PM: Value: 4
1:48:07 PM: Value: 9
1:48:07 PM: Value: 16
1:48:08 PM: Value: 25
1:48:08 PM: Value: 36
1:48:09 PM: Value: 49
1:48:09 PM: Value: 64
1:48:10 PM: Value: 81
1:48:10 PM: Value: 100
```

**Comment by chyczewski-maciej** [comment]
@bender2k14 
It's not really a problem when you have 10 elements (tasks started at the same time). 
It becomes a problem when you have thousands of elements and you make a call to a separate service for each of them. If you start all those tasks at the same time you'll bomb the service which will result in cutting you off or taking down the service.

What I'd expect is one of the following behaviors (where I believe the first one is the right approach):
- Sequence is enumerated lazily and sequentially, as the action passed to `IterT` is being executed for each of the elements
- Sequence is enumerated and actions are executed in parallel. This would most likely require a parameter to control the degree of parallelism.

Currently it's a mix of those two. Tasks are enumerated eagerly (before they are needed) while the action is executed sequentially. Since enumeration might be way faster than execution of the action it might lead to some kind of back pressure where you have millions of tasks started (and/or finished) waiting for the action to be executed for them.

### Sequential manner 
Here the next element is not requested until action is executed for the previous one. 

Optionally this could also have a small buffer that would allow to start a *limited* number of tasks in advance.

Simplified implementation
```
public static async Task IterTSequentially<T>(this IEnumerable<Task<T>> elements, Action<T> action)
{
    foreach (var item in elements)
        action(await item);
}
```

Expected output
```
21:16:38: Starting the application
21:16:38: Starting GetNumberAsync Task #1
21:16:39: Value: 1
21:16:39: Starting GetNumberAsync Task #2
21:16:40: Value: 4
21:16:41: Starting GetNumberAsync Task #3
21:16:42: Value: 9
21:16:42: Starting GetNumberAsync Task #4
21:16:43: Value: 16
21:16:44: Starting GetNumberAsync Task #5
21:16:45: Value: 25
21:16:45: Starting GetNumberAsync Task #6
21:16:46: Value: 36
21:16:47: Starting GetNumberAsync Task #7
21:16:48: Value: 49
21:16:48: Starting GetNumberAsync Task #8
21:16:49: Value: 64
21:16:50: Starting GetNumberAsync Task #9
21:16:51: Value: 81
21:16:51: Starting GetNumberAsync Task #10
21:16:52: Value: 100
```


### Parallel manner

Simplified implementation. 

It should take a parameter that'd control the degree of parallelism but let's skip it for simplicity.
```
public static async Task IterTParallel<T>(this IEnumerable<Task<T>> elements, Action<T> action)
{
    await Task.WhenAll(elements.Map(async item => action(await item)));
}
```

Expected output
In this case the action is executed in parallel for each element.
In this `action` will be executed in parallel.
```
21:18:53: Starting the application
21:18:53: Starting GetNumberAsync Task #1
21:18:53: Starting GetNumberAsync Task #2
21:18:53: Starting GetNumberAsync Task #3
21:18:53: Starting GetNumberAsync Task #4
21:18:53: Starting GetNumberAsync Task #5
21:18:53: Starting GetNumberAsync Task #6
21:18:53: Starting GetNumberAsync Task #7
21:18:53: Starting GetNumberAsync Task #8
21:18:53: Starting GetNumberAsync Task #9
21:18:53: Starting GetNumberAsync Task #10
21:18:54: Value: 1
21:18:54: Value: 4
21:18:54: Value: 9
21:18:54: Value: 36
21:18:55: Value: 25
21:18:55: Value: 81
21:18:55: Value: 49
21:18:55: Value: 100
21:18:55: Value: 64
21:18:55: Value: 16
```



In case it's still not clear I'll try to come up with better examples tomorrow.

**Comment by chyczewski-maciej** [comment]
Here's another drawback of current implementation. 
It does not stop the enumeration (or creation of the tasks in following example) of the `IEnumerable<Task<T>>` whenever the first failed task is encountered but it does stop executing actions for those results. Meaning: plenty of unnecessary tasks might be started and executed.

`Does_not_enumerate_the_sequence_after_a_failed_task` shows the behavior of current implementation. 

`Does_not_enumerate_the_sequence_after_a_failed_task_when_IterT_is_sequential` shows how it would behave if `IterT` would enumerate (start) tasks lazily.

```
public static class EnumerableTExtensions
{
    public static async Task IterTSequentially<T>(this IEnumerable<Task<T>> elements, Action<T> action)
    {
        foreach (var item in elements)
            action(await item);
    }
}

public class TestClass
{
    [Fact]
    public async Task Does_not_enumerate_the_sequence_after_a_failed_task()
    {
        await Assert.ThrowsAsync<Exception>(async () => await GetNumbersAsync().IterT(value => Thread.Sleep(500)));
    }

    [Fact]
    public async Task Does_not_enumerate_the_sequence_after_a_failed_task_when_IterT_is_sequential()
    {
        await Assert.ThrowsAsync<Exception>(async () => await GetNumbersAsync().IterTSequentially(value => Thread.Sleep(500)));
    }

    private IEnumerable<Task<int>> GetNumbersAsync()
    {
        yield return Task.FromResult(0);
        yield return Task.FromResult(1);
        yield return Task.FromException<int>(new Exception());
        Assert.True(false, "It should not enumerate further after a failed was encountered");
        yield return Task.FromResult(3);
    }
}
```

**Comment by bender2k14** [comment]
Ok, I think I understand now.

I am with @chyczewski-maciej in that the first implementation I think of for this `IterT` is a "sequential" one:
```
public static async Task<Unit> IterT< A>(
    this IEnumerable<Task<A>> mma,
    Action<A> f
) {
    foreach (var ma in mma) {
        var a = await ma;
        f(a);
    }
    return unit;
}
```

The current implementation is not equivalent to this.

> Here's another drawback of the current implementation.
> It does not stop the enumeration (or creation of the tasks in following example) of the IEnumerable<Task<T>> whenever the first failed task is encountered but it does stop executing actions for those results. Meaning: plenty of unnecessary tasks might be started and executed.

Yes, I see what you mean.  However, I don't quite follow your tests for that.  Here is my attempt to demonstrate this undesirable behavior.

```
[Fact]
public async Task IterT_SomeTaskFaulted_NoSubsequentTasksCreated() {
    var throwIndex = 3;
    var maxIndex = 0;
    var mma = Enumerable
        .Range(1, 9)
        .Map(async i => {
            await Task.Delay(1);
            maxIndex = Math.Max(i, maxIndex);
            if (i == throwIndex)
                throw new Exception();
            return i;
        });

    var e = await Record.ExceptionAsync(() => mma.IterT(_ => { }));

    Assert.NotNull(e);
    Assert.Equal(throwIndex, maxIndex); // fails here because max == 9
}
```



@louthy, do you understand now?  What is your intended behavior for this `IterT`?  For clarity, could you reply with a very simple / concrete / straightforward implementation of this `IterT` that matches your intended behavior?

**Comment by louthy** [comment]
I think there's a fundamental misunderstanding here.  `Iter` and `IterT` isn't there to pass judgement about how much work you want to do.  `IterT` especially is just an auto-generated function that wraps a nested fold over the inner and outer types, there's no special knowledge that the code has about your intentions.  It just iterates every outer value and for each one it iterates every inner value.

If you want some smarts in there, you need to write it.  Any smarts I decide upon will almost certainly be wrong for someone else.

**Comment by gwintering** [comment]
@chyczewski-maciej - First let me say this is an interesting discussion you've initiated with several subtle considerations.  I both agree and disagree with various parts and will try to separate them here.

### Should `IterT` stream tasks or request at batch of them from the generator?

Here I agree with you:  `IterT` should stream the tasks instead of requesting a batch of them.  So the pattern should be
```
request task 1 generation
operate on task 1
request task 2 generation
operate on task 2
...
```
It appears that right now we first get the batch of tasks then work on them like
```
request task 1 generation
request task 2 generation
...
operate on task 1
operate on task 2
...
```
I've intentionally left the nature of `operate` vague because I think that is the next consideration.

@bender2k14  -I think you're interpreting it right; the complain is that generation and operation are not interleaved.  @louthy I think this is a bug, do you agree?  Don't we want interleaving because it leverages the coroutine nature of the generator? 

### Should the operation on the task await completion before requesting generation of the next task?

No.  In most cases, exploiting concurrency is the entire point of using tasks.  I personally have been able to reduce execution time by several magnitudes by employing `IterT` in it's present form.  

If interleaving is fixed/respected, it's easy to add any form of awaiting/queuing/throttling by composing/decorating the action.  However, if awaiting is baked into `IterT`, there's no easy way to recover concurrent execution (while `Traverse` would work, that confusingly crosses the command/side effect vs query/return value divide).

### Concurrency vs Parallelism

Most of the time we use tasks for non blocking IO on a single thread instead of CPU bound work on several threads.  When doing non blocking IO, ramping the CPU up is a sign of success, not failure, because it means you're program is continuing to do useful work instead of waiting for IO. 

With non blocking IO,  "Uncontrolled parallelism" is not an issue because we're locally using a single thread, even though we may have numerous in progress tasks that represent the work of outside processes.

When actually doing CPU bound work, the TPL already gives us tools like `Parallel.ForAll` to control the degree of parallelism.

>If  you want some smarts in there, you need to write it.  Any smarts I  decide upon will almost certainly be wrong for someone else.

Agree @louthy , @chyczewski-maciej is worried about the capacity and DOS threshold of a particular API, which is an example of needing to limit the amount of non blocking IO.  But whether any form of awaiting/queuing/throttling is necessary, and how it should be implemented, is fundamentally extrinsic and context specific so it's probably not something that belongs in this library.

### Possible options

While I don't think we should introduce any form of queuing or rate limiting of tasks, we could form a new task monad that composes task in the wait-for-the-last-one manner that @chyczewski-maciej  desires.  All the methods could delegate to `MTask` except for `Plus`, which would look like this:
```csharp
[Pure]
public async Task<A> Plus(Task<A> ma, Task<A> mb)
{
    await ma;
    return await mb;
}
```
I haven't tested this, but I think it would make `IterT` behave such that only one task was ever active.

@chyczewski-maciej  another option, which I've used in the past when in a similar situation, was to put [echo-process](https://github.com/louthy/echo-process) in front of the API.  Then you can use actors to both control rate limiting and provide sophisticated failure recovery behavior (retries etc.).

**Comment by louthy** [comment]
@gwintering That's a good point about `MTask.Plus` - this would do the sequential operation:
```c#
await GetNumbersAsync().Fold(0.AsTask(), (s, t) => default(MTask<int>).Plus(s, t));
```

**Comment by louthy** [comment]
I haven't tested it, but a windowed iteration might look like this:
```c#
public static async Task<Unit> WindowedIter<A>(this IEnumerable<Task<A>> ma, int windowSize)
{
    var iter = ma.GetEnumerator();
    var endStream = false;

    IEnumerable<Task<A>> YieldWindow()
    {
        for(int i = 0; i < windowSize && (endStream = iter.MoveNext()); i++)
        {
            yield return iter.Current;
        }
    }

    while (!endStream)
    {
        await Task.WhenAll(YieldWindow());
    }
    return default;
}
```
Not super pretty, but I think should work.

**Comment by gwintering** [comment]
@louthy I don't think we can't use `MTask<>.Plus` because it awaits the second task only if the first task fails.  I was proposing something like `MTaskSingly<>` that would await both tasks.

**Comment by louthy** [comment]
@gwintering I've added `MTaskSerial` which does as you suggest.  I have also added `WindowIter`, but it has a different implementation to the code above, and instead builds `N` streams of tasks.  i.e.
```c#
   tasks.WindowedIter(10, f);
```
Would run 10 _streams_, where each stream would take an item from the list of tasks, await it, then got and get another.  And so they have a capacity that is constantly in use until all items in the lazy stream are exhausted.

The code I posted yesterday would have to await 10 tasks completing before moving onto the next 10 tasks, and so this will improve throughput, but also keep it constrained.

It can also run like so:
```c#
   tasks.WindowedIter(f);
```
Where the window size is `Environment.ProcessorCount / 2`

This is all in `v3.3.12`

**Comment by chyczewski-maciej** [comment]
Thanks guys for your input. The issue for me was that `IterT` on `IEnumerable<Task<T>>` behaved in a way I'd never expect. 

Is there a way to prevent other people from assuming it'd iterate elements of the sequence as they are needed (without calculating any of them ahead? Of course if you believe there's a need for it

**Comment by louthy** [comment]
> Is there a way to prevent other people from assuming it'd iterate elements of the sequence as they are needed 

I'm not sure there's any need.  The behavior is a nested fold with a unit state:
```c#
    IEnumerable<Task<Foo>> tasks;
    tasks.Fold(unit, (_, t) => t.Fold(unit, f));
```
`Iter` is about side-effects, it's unknown what those side-effects are going to be.  If the tasks are truly using IO then who's to say how much throughput any one machine could do in parallel?  The .NET scheduler may not be as effective as you'd like in all situations, but `IterT` can't make that judgement.  It's not waiting for results, it's firing an forgetting - which is kind of implied by the fact that `Iter` returns `Unit`.  And so it will try to generate as many tasks as it can and let the .NET scheduler deal with it.

If it did anything else then I'd argue it wouldn't be iteration.  It has no mandate to slow down, or do cunning stuff like wait for `n` tasks to complete before moving on.  

`WindowIter` is explicit in its behaviour and I think that's the correct approach.

**Comment by chyczewski-maciej** [comment]
@louthy, @gwintering, @bender2k14 
Thanks for the effort you guys put into it (and the entire library - it rises C# to a new level ) :)
